### PR TITLE
Introduce lowbit quantized linear MPS kernels

### PR DESCRIPTION
**Pull request by manuelcandales**
Summary:
The following is the directory structure of the submitted code under torchao

```
experimental/
├── kernels/
│   └── mps/
│       ├── codegen/
│       │   └── gen_metal_shader_lib.py
│       ├── metal/
│       │   └── (metal shaders)
│       ├── src/
│       │   └── (tensor agnostic mps kernel implementations)
│       ├── test/
│       │   └── (directly test mps kernel implementations)
│       └── metal.yaml
└── ops/
    └── mps/
        ├── register.mm
        ├── setup.py
        └── test/
            └── (test torch custom ops)
```

Differential Revision: D63342895

### CODE REVIEW

File: torchao/experimental/ops/mps/register.mm [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    const Tensor& B,
+    int64_t group_size,
+    const Tensor& SZ) {
+  auto M = A.size(0);
+  auto N = B.size(0);
+  auto K = A.size(1);
+  auto C = at::empty({M, N}, A.options());
+
+  LowBitQuantWeights<nbit>::linear(
+      getMTLBufferStorage(A),
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/register.mm:35]: Do we not need to assert that tensor is already on metal device?

For example if subsequent kernel is on cpu, then it will assume that the data is already available in the buffer pointed by the pointer. But depending on whether the work is submitted _and_ finished on GPU this may or may not be true.

e.g. Here C is created on same device as A.

File: torchao/experimental/ops/mps/register.mm [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  return C;
+}
+
+// LowBit Packing on CPU Backend
+template <int nbit>
+Tensor pack_weights_cpu_kernel(const Tensor& W) {
+  auto N = W.size(0);
+  auto K = W.size(1);
+  auto B = at::empty({N, nbit * K / 8}, W.options());
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/register.mm:53]: You need various asserts for different bit-width that make sure the K and N dim are of values that align with packing requirement.

For future variant we also may have to make sure to allow for padding but that can come later

File: torchao/experimental/ops/mps/register.mm [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  uint8_t* w_ptr = W.data_ptr<uint8_t>();
+  uint8_t* b_ptr = B.data_ptr<uint8_t>();
+
+  LowBitQuantWeights<nbit>::pack(w_ptr, b_ptr, N, K);
+
+  return B;
+}
+
+// Registers _C as a Python extension module.
+PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {}
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/register.mm:64]: Why do you need this?

File: torchao/experimental/ops/mps/setup.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/setup.py:1]: why this file at all?

File: torchao/experimental/ops/mps/test/test_lowbit.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class TestLowBitQuantWeightsLinear(unittest.TestCase):
+    cases = [
+        (nbit, *param)
+        for nbit in range(1, 8)
+        for param in [
+            (1, 32, 32, 32),
+            (128, 48, 64, 32),
+            (17, 1024, 512, 256),
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/test/test_lowbit.py:31]: definitely need more corner cases.

File: torchao/experimental/ops/mps/test/test_lowbit.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        ]
+    ]
+
+    def _init_tensors(self, group_size, M, K, N, nbit, device="mps"):
+        max_abs = 1 << (nbit - 1)
+        ceil_K_group_size = (K + group_size - 1) // group_size
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/test/test_lowbit.py:37]: This is making it sound like you can handle any groupsize but tests dont account for it

File: torchao/experimental/ops/mps/test/test_lowbit.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        # SZ is (K // group_size, N, 2)
+        N = W.shape[0]
+        K = W.shape[1]
+        max_abs = 1 << (nbit - 1)
+        W = W.to(torch.float32) - max_abs
+        scales = (
+            SZ[:, :, 0].t().unsqueeze(2).repeat(1, 1, group_size).view(N, -1)[:, :K]
+        )
+        zeros = SZ[:, :, 1].t().unsqueeze(2).repeat(1, 1, group_size).view(N, -1)[:, :K]
+        W = scales * W + zeros
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/test/test_lowbit.py:63]: I like this for reference impl

File: torchao/experimental/ops/mps/test/test_lowbit.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return wrapper
+
+    return decorator
+
+
+class TestLowBitQuantWeightsLinear(unittest.TestCase):
+    cases = [
+        (nbit, *param)
+        for nbit in range(1, 8)
+        for param in [
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/test/test_lowbit.py:28]: indentation fix needed?

File: torchao/experimental/ops/mps/test/test_lowbit.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        A, W, SZ = self._init_tensors(group_size, M, K, N, nbit=nbit)
+        packing_op = getattr(torch.ops.torchao, f"_pack_weight_{nbit}bit")
+        linear_op = getattr(torch.ops.torchao, f"_linear_fp_act_{nbit}bit_weight")
+        B = packing_op(W.cpu()).to("mps")
+        result = linear_op(A, B, group_size, SZ).cpu()
+        expected = self._reference_linear_lowbit_quant_weights(
+            A.cpu(), W.cpu(), group_size, SZ.cpu(), nbit=nbit
+        )
+        max_err = torch.max(torch.abs(result - expected))
+        self.assertLess(max_err, 0.001)
```
Comment (kimishpatel) [inline torchao/experimental/ops/mps/test/test_lowbit.py:78]: Use torch.testing.assert_close

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left quite a few comments. Lets address those. At high level

1. Figure out plan to how we gonna read .metal file. Likely similar to what torch does but see what MLX does too
2. Better test cases

File: torchao/experimental/ops/mps/test/test_lowbit.py (hunk 8 above)
Comment (manuelcandales) [reply torchao/experimental/ops/mps/test/test_lowbit.py:28]: this was the linter's output

File: torchao/experimental/kernels/mps/metal/divbit.metal [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+    constexpr uint8_t zero_shift = 1 << (nbit - 1);
+    constexpr uint8_t inv_nbit = 8 / nbit;
+    constexpr uint8_t minimask = (1 << nbit) - 1;
+
+    float rc = 0.0;
+    uint k = 0;
+    for (uint32_t kb = 0; kb < k_block ; kb ++) {
+      const T scale = scalesAndZeros[(kb * N + n) * 2 + 0];
+      const T zero = scalesAndZeros[(kb * N + n) * 2 + 1] - scale * T(zero_shift);
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/divbit.metal:30]: My intent was to be consistent with PyTorch's int4mm operator. Don't we want that?

File: torchao/experimental/kernels/mps/metal/int4mm.metal [hunk 11]
```diff
@@ -0,0 +1,80 @@
+#include <metal_stdlib>
+using namespace metal;
+
+// dispatchThreads:MTLSizeMake(N, M, 1)
+
+kernel void weight_to_int4pack(constant int *W [[buffer(0)]],
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/int4mm.metal:6]: It isn't. My intent was to keep all version of the shaders in the metal folder. But I am deleting them to make this PR cleaner.

File: torchao/experimental/kernels/mps/metal/int4mm.metal [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  const uint K = sizes.y;
+  const uint n = thread_index.x; // 0..N-1
+  const uint k2 = thread_index.y; // 0..K/2-1
+  int32_t src_val0 = W[n * K + 2 * k2];
+  int32_t src_val1 = W[n * K + 2 * k2 + 1];
+  outputData[n * (K / 2) + k2] = (uint8_t(src_val1) << 4) | uint8_t(src_val0);
+}
+
+template<typename T, unsigned groupSize>
+kernel void int4pack_mm(
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/int4mm.metal:20]: No, we don't need any of the int1, int2 or int4. I left them there, because I wrote them first, and then I wrote divbit. And like I said earlier, my intent was to not delete any shaders from the metal folder, even after they get superseded by an improved version. But will delete to make things cleaner

File: torchao/experimental/kernels/mps/metal/int3mm_vec.metal [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        uchar4 w_vals1 = uchar4(w_val4, w_val5, w_val6, w_val7);
+
+        rc += dot(a_vals0, scale * float4(w_vals0) + zero);
+        rc += dot(a_vals1, scale * float4(w_vals1) + zero);
+      }
+    }
+    outputData[m * N + n] = T(rc);
+}
+
+#define INSTANTIATE_INT3MM(DTYPE, GSIZE)                                 \
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/int3mm_vec.metal:73]: I will delete this shader too, to make the PR cleaner

File: torchao/experimental/kernels/mps/metal/divbit.metal (hunk 10 above)
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/divbit.metal:30]: My intent was to make these ops consistent with the current PyTorch implementation of int4

File: torchao/experimental/ops/mps/register.mm (hunk 3 above)
Comment (manuelcandales) [reply torchao/experimental/ops/mps/register.mm:64]: I believe this piece is what allows me to do: `import torchao_mps_ops`, instead of having to load the library path

File: torchao/experimental/ops/mps/setup.py (hunk 4 above)
Comment (manuelcandales) [reply torchao/experimental/ops/mps/setup.py:1]: This allows me to do:
`pip install .`
in order to install the custom library

File: torchao/experimental/kernels/mps/metal/divbit.metal [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+      for(uint idx = 0; idx < groupSize && k < K; idx++, k++) {
+        const auto a_val = float(A_ptr[k]);
+        uint8_t b_val = B_ptr[(n * K + k) / inv_nbit];
+        uint8_t shift = nbit * (k % inv_nbit);
+        uint8_t mask = minimask << shift;
+        b_val = (b_val & mask) >> shift;
+        rc += a_val * float(scale * T(b_val) + zero);
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/metal/divbit.metal:37]: K doesn't need to be a multiple of groupsize. That's why `idx` and `k` are looped together. You asked earlier where is the tail handled. This is where it is handled.

File: torchao/experimental/kernels/mps/src/lowbit.h [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+      [computeEncoder setComputePipelineState:cpl];
+      [computeEncoder setBuffer:a_buf offset:0 atIndex:0];
+      [computeEncoder setBuffer:b_buf offset:0 atIndex:1];
+      [computeEncoder setBuffer:sz_buf offset:0 atIndex:2];
+      [computeEncoder setBuffer:out_buf offset:0 atIndex:3];
+      [computeEncoder setBytes:sizes.data()
+                        length:sizeof(uint32_t) * sizes.size()
+                       atIndex:4];
+      dispatch_fn(computeEncoder, maxThreadsPerGroup, M, N, K);
+      finalize_block(mpsStream);
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/lowbit.h:140]: yeah, the ATen path uses torch's MPSStream and torch's utilities, which handle committing to command buffer and all that.

File: torchao/experimental/kernels/mps/src/lowbit.h [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+#include <torchao/experimental/kernels/mps/src/dispatch.h>
+#include <torchao/experimental/kernels/mps/src/packing.h>
+
+#include <fstream>
+#include <sstream>
+
+#ifdef ATEN
+#include <ATen/native/mps/OperationUtils.h>
+using namespace at::native::mps;
+inline void finalize_block(MPSStream* mpsStream) {}
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/lowbit.h:21]: this is to have minimal standalone code that doesn't depend on torch

File: torchao/experimental/kernels/mps/test/test_lowbit.mm [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+template <>
+const std::string_view type_string<float>() {
+  return "float";
+}
+template <>
+const std::string_view type_string<Float16>() {
+  return "half";
+}
+
+inline id<MTLBuffer> allocSharedBuffer(id<MTLDevice> device, unsigned length) {
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/test/test_lowbit.mm:30]: No, because the actual kernel doesn't need this. allocShaderBuffer is only used for testing

File: torchao/experimental/ops/mps/test/test_lowbit.py (hunk 6 above)
Comment (manuelcandales) [reply torchao/experimental/ops/mps/test/test_lowbit.py:37]: We could handle any groupsize. But in practice we only handle the ones instantiated in the shaders: 32, 64, 128, 256. This also agrees with PyTorch's int4 implementation

File: torchao/experimental/ops/mps/test/test_lowbit.py (hunk 5 above)
Comment (manuelcandales) [reply torchao/experimental/ops/mps/test/test_lowbit.py:31]: Adding them...

File: torchao/experimental/kernels/mps/test/test_lowbit.mm [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+} // namespace lowbit
+} // namespace mps
+} // namespace kernels
+} // namespace torchao
+
+template <typename T, int nbit>
+void run_test() {
+  torchao::kernels::mps::lowbit::LowBitTester<T, nbit> tester(
+      256, 256, 256, 32);
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/test/test_lowbit.mm:146]: Adding more tests cases...

File: torchao/experimental/kernels/mps/test/bfloat16.h [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * This implementation is copied from
+ * executorch/runtime/core/portable_type/bfloat16.h
+ */
+
+inline float f32_from_bits(uint16_t src) {
+  float res = 0;
+  uint32_t tmp = src;
+  tmp <<= 16;
+  std::memcpy(&res, &tmp, sizeof(tmp));
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/test/bfloat16.h:20]: std::bit_cast is C++20, but IIRC ExecuTorch is C++17

File: torchao/experimental/kernels/mps/test/bfloat16.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+  return res;
+}
+
+inline uint16_t bits_from_f32(float src) {
+  uint32_t res = 0;
+  std::memcpy(&res, &src, sizeof(res));
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/test/bfloat16.h:26]: see above

File: torchao/experimental/kernels/mps/src/lowbit.h [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+    id<MTLBuffer> a_buf,
+    id<MTLBuffer> b_buf,
+    int64_t qGroupSize,
+    id<MTLBuffer> sz_buf,
+    id<MTLBuffer> out_buf,
+    int32_t M,
+    int32_t K,
+    int32_t N,
+    const std::string_view type_str) {
+  const std::string shader_path = std::string(METAL_SHADER_DIR) + "/" +
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/lowbit.h:154]: fixed

File: torchao/experimental/kernels/mps/src/lowbit.h [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+      static_cast<uint32_t>(M),
+      static_cast<uint32_t>(K),
+      static_cast<uint32_t>(N),
+      0};
+
+  MPSStream* mpsStream = getCurrentMPSStream();
+  dispatch_sync_with_rethrow(mpsStream->queue(), ^() {
+    @autoreleasepool {
+      id<MTLComputeCommandEncoder> computeEncoder = mpsStream->commandEncoder();
+      MetalShaderLibrary lib = compileLibraryFromFile(shader_path);
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/lowbit.h:124]: fixed

File: torchao/experimental/kernels/mps/src/lowbit.h [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+#endif
+
+namespace torchao {
+namespace kernels {
+namespace mps {
+namespace lowbit {
+namespace {
+
+static constexpr std::string_view METAL_SHADER_DIR =
+    "/Users/mcandales/fbsource/xplat/pytorch/ao/torchao/experimental/kernels/mps/metal";
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/lowbit.h:33]: fixed

File: torchao/experimental/kernels/mps/src/OperationUtils.h [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+// This source code is licensed under the license found in the
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <iostream>
+
+static void fail(const std::string& str) {
+  std::cerr << str << std::endl;
+  abort();
```
Comment (malfet) [inline torchao/experimental/kernels/mps/src/OperationUtils.h:13]: This looks more C++ to me.. (I.e. callee can catch the error if need be)
```suggestion
  throw std::runtime_error(str);
```

File: torchao/experimental/kernels/mps/src/OperationUtils.h [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+  (void)queue;
+  block();
+}
+
+inline id<MTLDevice> getMetalDevice() {
+  NSArray* devices = [MTLCopyAllDevices() autorelease];
+  if (devices.count == 0) {
+    fail("Metal is not supported");
+  }
+  return devices[0];
```
Comment (malfet) [inline torchao/experimental/kernels/mps/src/OperationUtils.h:33]: Consider adding autorelease here. Also do you compile with ARC or without it?
```suggestion
  @autorelese {
    NSArray* devices = [MTLCopyAllDevices() autorelease];
    if (devices.count == 0) {
      fail("Metal is not supported");
    }
    return devices[0];
  }
```

File: torchao/experimental/kernels/mps/src/OperationUtils.h [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+static id<MTLDevice> MTL_DEVICE = getMetalDevice();
+
+static id<MTLLibrary> compileLibraryFromSource(
+    id<MTLDevice> device,
+    const std::string& source) {
+  NSError* error = nil;
+  MTLCompileOptions* options = [MTLCompileOptions new];
+  [options setLanguageVersion:MTLLanguageVersion3_1];
```
Comment (malfet) [inline torchao/experimental/kernels/mps/src/OperationUtils.h:43]: You probably want to disable fast math here as well...

File: torchao/experimental/kernels/mps/src/dispatch.h [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+
+#pragma once
+
+#include <functional>
+
+namespace torchao {
+namespace kernels {
+namespace mps {
+namespace lowbit {
+namespace dispatch {
```
Comment (malfet) [inline torchao/experimental/kernels/mps/src/dispatch.h:15]: Nit (why not use C++17 conventions)
```suggestion
namespace torchao::kernels::mps::lowbit::dispatch {
```

File: torchao/experimental/kernels/mps/src/dispatch.h [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+
+inline void dispatch_mm(
+    id<MTLComputeCommandEncoder> encoder,
+    int32_t maxThreadsPerGroup,
+    int32_t M,
+    int32_t N,
+    int32_t K) {
+  (void)K;
```
Comment (malfet) [inline torchao/experimental/kernels/mps/src/dispatch.h:23]: Or you can use good old `__attribute__((unused))` if you want to stick to C++14 
```suggestion
    [[maybe_unused]]  int32_t K) {
```

File: torchao/experimental/ops/mps/setup.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+from setuptools import setup
+from torch.utils.cpp_extension import CppExtension, BuildExtension
+
+setup(
+    name="torchao_mps_ops",
+    version="1.0",
+    ext_modules=[
+        CppExtension(
+            name="torchao_mps_ops",
+            sources=["register.mm"],
```
Comment (malfet) [inline torchao/experimental/ops/mps/setup.py:17]: I think you either need to package your kernels into the wheel, or using `bin2c` just convert them to headers. Or convert them into a metallib

File: torchao/experimental/kernels/mps/src/OperationUtils.h [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+// LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <iostream>
+
+static void fail(const std::string& str) {
+  std::cerr << str << std::endl;
+  abort();
+}
```
[the remark this answers is not in the document]
Comment (manuelcandales) [reply torchao/experimental/kernels/mps/src/OperationUtils.h:14]: throwing exception now
